File: src/easy/add-binary.ts
export const addBinary = (a: string, b: string): string => {
    let i = a.length - 1;
    let j = b.length - 1;
    let carry = 0;
    const target = Array(Math.max(a.length, b.length));
		let k = target.length;
    while (i >= 0 || j >= 0 || carry > 0) {
        let sum = carry;
			
				if (i >= 0) sum += a.charCodeAt(i) - 48;
        if (j >= 0) sum += b.charCodeAt(j) - 48;

				target[k] = sum % 2
        carry = Math.floor(sum / 2);
				k--
        i--;
        j--;
    }

			return target.join("");
};
 

File: src/easy/add_binary.zig
const std = @import("std");

const expectEqualSlices = std.testing.expectEqualSlices;
const MathTools = @import("./math_tools.zig");

const Allocator = std.mem.Allocator;

pub fn addBinary(allocator: Allocator, a: []const u8, b: []const u8) ?[]const u8 {
    const max = MathTools.maxValue(usize, a.len, b.len);
    var target = allocator.alloc(u8, max + 1) catch return null;
    var i = a.len;
    var j = b.len;
    var k = target.len;
    var carry: u8 = 0;

    while (i > 0 or j > 0 or carry > 0) {
        var sum: u8 = carry;

        if (i > 0) {
            i -= 1;
            sum += MathTools.asciiAsDigit(a[i]);
        }
        if (j > 0) {
            j -= 1;
            sum += MathTools.asciiAsDigit(b[j]);
        }

        k -= 1;
        target[k] = MathTools.digitAsAscii(@mod(sum, 2));
        carry = @divFloor(sum, 2);
    }
    return target[k..];
}

test "adding 11 to 1 equals 100" {
    const a = "11";
    const b = "1";
    const expected = "100";
    const allocator = std.testing.allocator;
    const result = addBinary(allocator, a, b);
    if (result) |actual| {
        try expectEqualSlices(u8, expected, actual);
        defer allocator.free(actual);
    } else return error.SkipZigTest;
}
test "adding 1010 to 1011 equals 10101" {
    const a = "1010";
    const b = "1011";
    const expected = "10101";
    const allocator = std.testing.allocator;
    const result = addBinary(allocator, a, b);
    if (result) |actual| {
        try expectEqualSlices(u8, expected, actual);
        defer allocator.free(actual);
    } else return error.SkipZigTest;
}
